File: packages/programming_using_c/grammar/words/js/token.ts
export class Token {
  private kind: string;
  private value: string;

  static createEmpty() {
    return new Token('', '');
  }
  static createNoun(v: string) {
    return new Token('n', v);
  }
  static createVerb(v: string) {
    return new Token('v', v);
  }
  static createPreposition(v: string) {
    return new Token('p', v);
  }
  static createEnd() {
    return new Token('.');
  }
  static createArticle() {
    return new Token('a', '');
  }

  private constructor(k: string, v: string = '') {
    this.kind = k;
    this.value = v;
  }

  equal(ch: string) {
    return this.kind === ch;
  }
  get() {
    return this.value;
  }
  isNoun() {
    return this.kind === 'n';
  }
  isVerb() {
    return this.kind === 'v';
  }
  isPreposition() {
    return this.kind === 'p';
  }
  isEnd() {
    return this.kind === '.';
  }
  isEmpty() {
    return this.kind === '';
  }
  isArticle() {
    return this.kind === 'a';
  }
  toString() {
    return `{k: ${this.kind}, v: ${this.value}}`;
  }
}

File: packages/programming_using_c/grammar/words/c++/token.cpp
#include "include/std_lib_facilities.h"

class Token
{
private:

public:
  string value;
  char kind;
  Token(char ch, string v) : kind(ch), value(v) {}

  static Token createNoun(string word);
  static Token createVerb(string word);
  static Token createEnd();
  static Token createExit();
  static Token createEmpty();
  static Token createArticle(string word);
  static Token createPreposition(string word);

  bool isEnd();
  bool isExit();
  bool isNoun();
  bool isVerb();
  bool isArticle();
  bool isPreposition();
  string get_value();
};

bool Token::isPreposition()
{
  return kind == 'p';
}
bool Token::isVerb()
{
  return kind == 'v';
}
bool Token::isNoun()
{
  return kind == 'n';
}
bool Token::isArticle()
{
  return kind == 'a';
}
bool Token::isExit()
{
  return kind == 'q';
}
bool Token::isEnd()
{
  return kind == '.';
}

Token Token::createEnd()
{
  return Token{'.', ""};
}
Token Token::createNoun(string word)
{
  return Token{'n', word};
}
Token Token::createExit()
{
  return Token{'q', ""};
}
Token Token::createVerb(string word)
{
  return Token{'v', word};
}
Token Token::createArticle(string word)
{
  return Token{'a', word};
}
Token Token::createPreposition(string word)
{
  return Token{'p', word};
}
Token Token::createEmpty()
{
  return Token{'0', ""};
}
string Token::get_value()
{
  return value;
}